lib/listener_manager.c

```diff
@@ -90,44 +93,137 @@ bool mptcpd_lm_listen(struct l_hashmap *lm,
-                return false;
+                return -1;
         }
 
         if (listen(fd, 0) == -1) {
                 l_error("Unable to listen on MPTCP socket: %s",
                         strerror(errno));
                 (void) close(fd);
+                return -1;
+        }
+
+        return fd;
+}
+
+static void close_listener(void *value)
+{
+        struct lm_value *const data = value;
+
+        /**
+         * @todo Should care if the @c close() call fails?  It
+         *       shouldn't fail since we only get here if the listener
+         *       socket was successfully opened earlier.
+         */
+        (void) close(data->fd);
+        l_free(data);
+}
+
+// ----------------------------------------------------------------------
+
+struct mptcpd_lm *mptcpd_lm_create(void)
+{
+        struct mptcpd_lm *lm = l_new(struct mptcpd_lm, 1);
+
+        // Map of IP address to MPTCP listener file descriptor.
+        lm->map  = l_hashmap_new();
+        lm->seed = l_getrandom_uint32();
+
+        if (!l_hashmap_set_hash_function(lm->map, mptcpd_hash_sockaddr)
+            || !l_hashmap_set_compare_function(lm->map,
+                                               mptcpd_hash_sockaddr_compare)
+            || !l_hashmap_set_key_copy_function(lm->map,
+						mptcpd_hash_sockaddr_key_copy)
+            || !l_hashmap_set_key_free_function(lm->map,
+                                                mptcpd_hash_sockaddr_key_free)) {
+                mptcpd_lm_destroy(lm);
+                lm = NULL;
+        }
+
+        return lm;
+}
+
+void mptcpd_lm_destroy(struct mptcpd_lm *lm)
+{
+        if (lm == NULL)
+                return;
+
+        l_hashmap_destroy(lm->map, close_listener);
+        l_free(lm);
+}
+
+bool mptcpd_lm_listen(struct mptcpd_lm *lm, struct sockaddr const *sa)
+{
+        if (lm == NULL || sa == NULL)
+                return false;
+
+        if (sa->sa_family != AF_INET && sa->sa_family != AF_INET6)
                 return false;
+
+        struct mptcpd_hash_sockaddr_key const key = {
+                .sa = sa, .seed = lm->seed
+        };
+
+        struct lm_value *data = l_hashmap_lookup(lm->map, &key);
+
+        /*
+          A listener already exists for the given IP address.
+          Increment the reference count.
+        */
+        if (data != NULL) {
+                data->refcnt++;
+
+                return true;
         }
 
-        if (!l_hashmap_insert(lm, L_UINT_TO_PTR(id), L_UINT_TO_PTR(fd))) {
-                l_error("Unable to map MPTCP address ID %u listener", id);
-                (void) close(fd);
+        data = l_new(struct lm_value, 1);
+
+        data->fd = open_listener(sa);
+        if (data->fd == -1) {
+                l_free(data);
+
+                return false;
+        }
+
+        if (!l_hashmap_insert(lm->map, &key, data)) {
+                l_error("Unable to map MPTCP address to listener.");
+
+                (void) close(data->fd);
+                l_free(data);
```

Comment:
These two lines could maybe use the `close_listener()` helper. Although that does involve throwing away some type safety because of the `void *` parameter. I ran in to this issue in `ndiffports` and made a "real" helper with the right parameter types, then a wrapper around that to fit the `void *` params required for ELL callbacks. Other solutions work too.